### PR TITLE
fix: refactor Input handling

### PR DESCRIPTION
**Pull request by fcolarich**
## What does this PR change?

What does this NOT do?! well for starters, it wont cuddle you in the middle of the night when you are feeling lonely.

What does DO do? Given the growing number of windows and menus and whatnot that we are adding, the input handling was getting out of control, particularly blocking and unblocking input. It would happen that a window would block input until closed and then another window would open blocking the same input, but when the previous window finishes closing after the second one loads, it blocks the input again. And all sorts of similar things. Making it REALLY difficult to ascertain which inputs should be enabled and when.
To avoid this, I piggybacked into a couple of existing systems that already handled input changes, but improved them so we can properly track how many times each input was requested to be blocked or unblocked. 
So, I expanded the `InputBlock `class so it can be used as intermediary between the already existing `InputMapComponent `and every Controller. Also I expanded the `InputMapComponent `so it has inside a dictionary tracking each input kind along with the number of block requests for it.
The `InputBlock `class can block several inputs at the same time, making use of the `InputMapComponent.Kind` Flag enum, but can be overriden to only block one input to save a bit of processing.
The `InputMapComponent `will only update its Active property when the block counter reaches either 1 or 0 (for disabling or enabling the inputs), in other cases it will just increase or decrease the appropriate counters.
Other than that, it was a matter of updating all places where input blocking/unblocking was done and make sure the number of blocks and unblocks was paired in all cases. This was time consuming for some parts of the game, but overall ended up improving some exit flows as well.
Also, added finally, the input blocking while the loading screen is showing, this way no input will be sent while displaying it.

Now, go find something to cuddle you in the middle of the night you brave soul!

## How to test the changes?

To test it just try to go through the normal flow of the game, chat messages, menus, jumping through worlds, etc. The input should behave as before and even better :) Now it should be impossible to move while teleporting and the loading screen is showing for example.
This fixes this issue as well: [https://github.com/decentraland/unity-explorer/issues/1793] As it was caused exactly by this kind of Input woes.
Also fixes an issue happening when opening the emote wheel and closing it quickly which sometimes would leave some inputs blocked

### CODE REVIEW

**Review by anicalbano — APPROVED** [review]
🟢 Tested thoroughly, everything appears to be working as expected

File: Explorer/Assets/DCL/Input/Component/InputMapComponent.cs
Comment (popuz) [inline Explorer/Assets/DCL/Input/Component/InputMapComponent.cs:53]: why it is error? and if so, can you describe it better inside the message

File: Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs
Comment (popuz) [inline Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs:4]: also not sure it is related to your change. probably can be removed

File: Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs
Comment (popuz) [inline Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs:81]: in this case I would keep it without brackets

File: Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs
Comment (popuz) [inline Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs:36]: could be probably `foreach`, which would be a bit more readable

File: Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs
Comment (popuz) [inline Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs:43]: maybe if you make it via paramas, then you can just iterate over it and take values directly from dictionary, instead of iterating dictionary. It will probably allow you to remove `singleValue` block of code. Same applies to `BlockInput` method

**Review by popuz — APPROVED** [review]
Nice refactoring 💪 
However, I would suggest some changes here and there :) While most of them are small and rather cosmetical, changing `BlockInput`/`UnblockInput ` from  `UnblockInputs(InputMapComponent.Kind kinds, bool singleValue = false)` to `UnblockInputs(params InputMapComponent.Kind[] kinds)` could be a nice improvement

File: Explorer/Assets/DCL/Input/Component/InputMapComponent.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Input/Component/InputMapComponent.cs:53]: AH its not error!! Was a log I was printing to validate stuff and this one escaped the cleanse!!

File: Explorer/Assets/DCL/Character/CharacterCamera/Systems/UpdateCameraInputSystem.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Character/CharacterCamera/Systems/UpdateCameraInputSystem.cs:93]: Oh I just moved things around here, those are not my magic numbers :P but I agree, I will put them into constants for now :)

File: Explorer/Assets/DCL/Backpack/BackpackSearchController.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Backpack/BackpackSearchController.cs:1]: not sure! but good catch, I will check!

File: Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Input/UnityInputSystem/Blocks/InputBlock.cs:43]: yeah I was thinking the same at some point :) seems like a good approach!

File: Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Input/Systems/ApplyInputMapsSystem.cs:4]: nice catch!!!

File: Explorer/Assets/DCL/Audio/UIAudioEventsBus.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Audio/UIAudioEventsBus.cs:6]: yeah, this was part of another change that in the end I didnt go through, but will be needed for the future :) Ill extract it

**Comment by fcolarich** [comment]
Will think of more optimizations for the block/unblock methods after merging :)

File: Explorer/Assets/DCL/Character/CharacterCamera/Systems/ControlCinemachineVirtualCameraSystem.cs
Comment (fcolarich) [reply Explorer/Assets/DCL/Character/CharacterCamera/Systems/ControlCinemachineVirtualCameraSystem.cs:132]: yep seems like a good idea :) I will look into this after merging with a couple of other optimizations we discussed

**Pull request merged**
